File: src/main/java/it/pagopa/ecommerce/eventdispatcher/mdcutilities/MDCContextLifterConfiguration.java
package it.pagopa.ecommerce.cdc.mdcutilities;

import jakarta.annotation.PostConstruct;
import jakarta.annotation.PreDestroy;
import org.springframework.context.annotation.Configuration;
import reactor.core.publisher.Hooks;
import reactor.core.publisher.Operators;

@Configuration
public class MDCContextLifterConfiguration {

    private final String mdcContextReactorKey = MDCContextLifterConfiguration.class.getName();

    @PostConstruct
    private void contextOperatorHook() {
        Hooks.onEachOperator(
                mdcContextReactorKey,
                Operators.lift(
                        (
                         scannable,
                         coreSubscriber
                        ) -> new MDCContextLifter<>(coreSubscriber)
                )
        );
    }

    @PreDestroy
    private void cleanupHook() {
        Hooks.resetOnEachOperator(mdcContextReactorKey);
    }
}

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/mdcutilities/MDCContextLifterConfiguration.kt
package it.pagopa.ecommerce.eventdispatcher.mdcutilities

import org.springframework.context.annotation.Configuration
import reactor.core.publisher.Hooks
import reactor.core.publisher.Operators
import javax.annotation.PostConstruct
import javax.annotation.PreDestroy

@Configuration
class MDCContextLifterConfiguration {

    private val mdcContextReactorKey = this::class.java.name

    @PostConstruct
    fun contextOperatorHook() {
        // register a hook applied to each operator in che reactive chain
        Hooks.onEachOperator(
            mdcContextReactorKey,
            // lift -> transforms a Subscriber to apply the MDC lifting functionality
            Operators.lift { _, coreSubscriber ->
                // for each subscriber, wrap it with the MDCContextLifter
                MDCContextLifter(coreSubscriber)
            },
        )
    }

    @PreDestroy
    fun cleanupHook() {
        // remove the hook to prevent memory leaks
        Hooks.resetOnEachOperator(mdcContextReactorKey)
    }
}
